### PR TITLE
Feat: Pro(b)log support

### PR DESCRIPTION
**Pull request by foxyseta**
Added Prolog and ProbLog parsers and queries. The two languages do not have proper scopes or injections, as far as I know (used the ISO standard to implement the parsers), so those queries are intentionally missing.

### CODE REVIEW

**Comment by foxyseta** [comment]
Not sure how to make sense of the failed pipelines logs.

**Comment by ribru17** [comment]
You need to add lockfile entries for the two parsers (not sure if this will fix all the tests)

**Comment by foxyseta** [comment]
Done!

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:1571]: No need to specify the branch if it's the default.

**Comment by clason** [comment]
Also, we expect to be able to download archives as with Github (cloning and checking out revisions is deprecated for security reasons). Does codeberg support that?

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:1571]:
```suggestion
```

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:1580]:
```suggestion
```

**Comment by clason** [comment]
> Not sure how to make sense of the failed pipelines logs.

Here's your issue: https://codeberg.org/foxy/tree-sitter-problog/src/commit/3687c8579bad69c0143852032f3830363f151312/grammar.js#L306

File: lua/nvim-treesitter/parsers.lua
Comment (foxyseta) [reply lua/nvim-treesitter/parsers.lua]: You are right, thank you! I got confused with the local install (from init.lua)!

**Comment by foxyseta** [comment]
> > Not sure how to make sense of the failed pipelines logs.
> 
> Here's your issue: https://codeberg.org/foxy/tree-sitter-problog/src/commit/3687c8579bad69c0143852032f3830363f151312/grammar.js#L306

Thanks! Fixed in the parser repo and updated the lockfile accordingly.

**Comment by clason** [comment]
You still might like to consider keeping both grammars in a single repository and inheriting from a common grammar, like tree-sitter-ocaml does.

**Comment by foxyseta** [comment]
> Also, we expect to be able to download archives as with Github (cloning and checking out revisions is deprecated for security reasons). Does codeberg support that?

Yes, you can do stuff like:

https://codeberg.org/foxy/tree-sitter-prolog/archive/main.zip
https://codeberg.org/foxy/tree-sitter-prolog/archive/main.tar.gz

And also

https://codeberg.org/foxy/tree-sitter-prolog/archive/main.bundle,

which is not supported by GitH*b.

**Comment by foxyseta** [comment]
> You still might like to consider keeping both grammars in a single repository and inheriting from a common grammar, like tree-sitter-ocaml does. I think this would be a better strategy.

All in the same branch? Is there any guide for grammar inheritance anywhere?

**Comment by clason** [comment]
> All in the same branch? Is there any guide for grammar inheritance anywhere?

Yes, just different directories. There's no official guidance (yet), but you can look at the upstream ocaml parser as an example.

> Yes, you can do stuff like:

I need `archive/<sha>.tar.gz`, but that probably works as well then.

> which is not supported by GitH*b.

🙄

**Comment by foxyseta** [comment]
> Yes, just different directories. There's no official guidance (yet), but you can look at the upstream ocaml parser as an example.

Okay, thank you! I can work on that soon and then do the relevant changes to this PR once I am ready.  
 
> I need `archive/<sha>.tar.gz`, but that probably works as well then.

Yes, that works too.

**Comment by foxyseta** [comment]
@clason Done! Hopefully I followed the example correctly.

File: queries/prolog/highlights.scm
Comment (ObserverOfTime) [inline queries/prolog/highlights.scm:9]:
```suggestion
((atom) @boolean
  (#any-of? @boolean "true" "false"))
```

File: queries/prolog/highlights.scm
Comment (ObserverOfTime) [inline queries/prolog/highlights.scm:18]:
```suggestion
(integer) @number
```

File: queries/prolog/highlights.scm
Comment (ObserverOfTime) [inline queries/prolog/highlights.scm:24]:
```suggestion
(float_number) @number.float
```

File: queries/prolog/highlights.scm
Comment (ObserverOfTime) [inline queries/prolog/highlights.scm:1]:
```suggestion
(comment) @comment @spell
```

File: queries/prolog/highlights.scm
Comment (ObserverOfTime) [reply queries/prolog/highlights.scm:18]: Oh I see.

Comment (foxyseta) [reply queries/prolog/highlights.scm:18]: @lucario387 are you talking about making `-` part of `integer` and removing the `negative_number` non-terminal altogether? Although I do agree that it is very weird to use `integer` to only refer to non-negative integer numbers, the current state of the parser is meant to reflect how the ISO specification defines the language. The currenty query is built keeping in mind we do not want to change the parser.

Comment (lucario387) [reply queries/prolog/highlights.scm:18]: Yes, that's what I meant. It's kindof weird for the ISO spec of a language to specifically denote negative number though. Is there any special usage for it?

Comment (foxyseta) [reply queries/prolog/highlights.scm:18]: I cannot find an explicit motivation anywhere, sorry. Could this has been done for unification/pattern matching, maybe? This way, `delta_is(-1)` can match `delta_is(-x)` for example. In the original specs, negative numbers are specified as:

```
6.3.1.2 Negative numbers
term = name, integer;
term = name, float number;
(Condition: name is "-")
```

So the non-terminal "name" is used, but then a side condition is specified.

I personally think sticking to the language itself is beneficial in this case, but I am still open to rewriting this part of the parser as long as I can document the change with a valid reason.

Comment (clason) [reply queries/prolog/highlights.scm:18]: In general, sticking too closely to the language spec is not a good idea; often a more pragmatic approach that focuses on the end goal (which for us is syntax highlighting) leads to a more robust and performant parser. [(Github says so.)](https://github.com/github/semantic/blob/3baf0b47ff2117a4724771687d458151ce2b92a7/docs/grammar-development-guide.md#rely-on-language-specs-but-not-too-much)

That being said, in the end it's your decision; this is not a critical blocker.

**Comment by ribru17** [comment]
Would you mind adding comment injections please?

File: queries/prolog/highlights.scm
Comment (ribru17) [inline queries/prolog/highlights.scm:38]: I think `":-"` and `"|"` should be added. Also `(comma)` should be given `@punctuation.delimiter` somewhere after this query

File: queries/prolog/highlights.scm
Comment (ribru17) [inline queries/prolog/highlights.scm:29]: 1100 (semicolon) should be moved to `@punctuation.delimiter`. To be honest I think most or all of the `operator_*` nodes should be prefixed with underscores in the grammar, these node names seem pretty unintuitive :sweat_smile:

Comment (foxyseta) [reply queries/prolog/highlights.scm:29]: Unlike periods, commas and semicolons are considered operators in Prolog, but I can see how thinking about them as punctuation would be more intuitive, so I'll change this.

I thought hiding nodes would prevent me from referring to them in queries?

File: queries/prolog/highlights.scm
Comment (foxyseta) [reply queries/prolog/highlights.scm:38]: I would maybe give `|` the same highlight group as its braces. Currently, only binary :- is colored as an operator, so I can color unary :- as one, too.

File: queries/prolog/highlights.scm
Comment (foxyseta) [reply queries/prolog/highlights.scm:18]: I understand! You make a valid point. I'll need some time to make the changes, and then ping you all when I am ready. Thanks.

File: queries/prolog/highlights.scm
Comment (ribru17) [reply queries/prolog/highlights.scm:29]: That would be a great compromise, thanks! Also if things like semicolons are meant to be operators then it's fine to keep it that way; my mistake :)

Comment (foxyseta) [reply queries/prolog/highlights.scm]: Whoopsie I hadn't seen this commit in time and committed already. It's okay, I like it better your way anyway.

**Comment by foxyseta** [comment]
I think I addressed all of the comments yesterday. Let me know if I misunderstood/forgot something!

File: queries/prolog/highlights.scm
Comment (ribru17) [inline queries/prolog/highlights.scm:18]: Thank you, I think this PR is almost good to go! One thing: 
```suggestion
  operator: _ @operator)
```
This way anonymous nodes can be captured. Also, would you mind updating the grammar (and then the lockfile) so that things like:
```js
        prec.left(
          1,
          seq(
            $._term,
            field(fields.operator, /(:\-)|(\-\->)/),
            $._term,
          ),
        ),
```
can be indexed? Sadly regexes become invisible in the grammar and thus not query-able. Something like the following would allow for operators that use regex rules to be highlighted:
```js
        prec.left(
          1,
          seq(
            $._term,
            field(fields.operator, alias(/(:\-)|(\-\->)/, $.arrow)),
            $._term,
          ),
        ),
```
(or you can of course replace `$.arrow` with something quoted if you prefer the node be anonymous, your call)

Comment (foxyseta) [reply queries/prolog/highlights.scm]: Should have fixed in https://github.com/nvim-treesitter/nvim-treesitter/pull/6810/commits/d95626829a0cda0c2b002dceb0745d6516ca37db

**Review by ribru17 — APPROVED** [review]

**Comment by clason** [comment]
@foxyseta please rebase on master (do not merge master into your branch!) and squash your commits

**Comment by lucario387** [comment]
woah woah, what's happening here, why's there all these commits? Did you forget to drop unrelated commits while interactive rebasing?

**Comment by foxyseta** [comment]
Apparently I did, sorry 😅  It's my first time dealing with unrelated commits during a rebase

**Comment by clason** [comment]
> > Yes, just different directories. There's no official guidance (yet), but you can look at the upstream ocaml parser as an example.
> 
> Okay, thank you! I can work on that soon and then do the relevant changes to this PR once I am ready.
> 
> > I need `archive/<sha>.tar.gz`, but that probably works as well then.
> 
> Yes, that works too.

Not the same way, though. This breaks the refactored installation on the `main` branch (which will become the default branch in the near future). There's a limit how many different codeforges we can (and want to) support.

**Comment by foxyseta** [comment]
I am sorry. Luckily, compatibility with Codeberg means compatibility with any ForgeJo (and maybe Gittea) instance. Are you considering rolling back this PR or adding compatibility to the `main` branch? Is there any way I can help?

**Comment by lucario387** [comment]
Having a GitHub/GitLab mirror of the repo will be nice (and probably required, unless there's enough parsers that's hosted on codeberg)

**Comment by foxyseta** [comment]
Yes, I had considered that option and it is definitely not a problem for me. Here you go:
https://github.com/foxyseta/tree-sitter-prolog. I'll make sure this mirror stays up to date. So discussion for ForgeJo support will open only if enough parsers are hosted on it? I usually push all my code to Codeberg, but I do not mind mirroring to GitHub my future parsers, too.

What precisely are the differences between Codeberg and GitHub breaking this? If it's an objective limitation of ForgeJo, maybe a patch could be submitted. If it's just different design choices between the two, I guess not.

**Comment by lucario387** [comment]
> What precisely are the differences between Codeberg and GitHub breaking this

We had bad experience with Codeberg before, such as flaky CIs due to codeberg being down at certain times. Another reason is different installation support, which `main` branch significantly reduced (on `master` there is support for makefile which was removed for `main` branch, for example).

**Comment by clason** [comment]
I already pushed a compat commit to `main`, but the flakiness of "smaller" codeforges (due to more frequent outages due to smaller resources) is a significant issue for our CI: if a test fails due to not being able to install a parser, our automated actions are blocked. So thank you for the mirror; I'll switch the parser to that.

**Comment by clason** [comment]
@foxyseta that doesn't look like a full mirror; it's missing the problog parser!

The difference is really small: the tarball for a commit (e.g., `https://codeforge.org/foxyseta/tree-sitter-prolog/<hash>.tar.gz`) when unpacked contains a `tree-sitter-prolog` directory with the repo contents, while Github and Gitlab append the hash to it (`tree-sitter-prolog-<hash>`).

**Comment by foxyseta** [comment]
Cool, thank you. Will make sure to also provide mirrors in the future.

**Pull request merged**
